chain/src/main/scala/org/bitcoins/chain/blockchain/Blockchain.scala

```diff
@@ -60,53 +61,57 @@ object Blockchain extends ChainVerificationLogger {
     *         we [[org.bitcoins.chain.blockchain.BlockchainUpdate.Successful successful]] connected the tip,
     *         or [[org.bitcoins.chain.blockchain.BlockchainUpdate.Failed Failed]] to connect to a tip
     */
-  def connectTip(header: BlockHeader, blockHeaderDAO: BlockHeaderDAO)(
+  def connectTip(
+      header: BlockHeader,
+      blockHeaderDAO: BlockHeaderDAO,
+      blockchains: Vector[Blockchain])(
       implicit ec: ExecutionContext,
       conf: ChainAppConfig): Future[BlockchainUpdate] = {
 
-    //get all competing chains we have
-    val blockchainsF: Future[Vector[Blockchain]] =
-      blockHeaderDAO.getBlockchains()
-
-    val tipResultF: Future[BlockchainUpdate] = blockchainsF.flatMap {
-      blockchains =>
-        val nested: Vector[Future[BlockchainUpdate]] = blockchains.map {
-          blockchain =>
-            val prevBlockHeaderOpt =
-              blockchain.find(_.hashBE == header.previousBlockHashBE)
-            prevBlockHeaderOpt match {
-              case None =>
-                logger.debug(
-                  s"No common ancestor found in the chain to connect to ${header.hashBE}")
-                val err = TipUpdateResult.BadPreviousBlockHash(header)
-                val failed = BlockchainUpdate.Failed(blockchain = blockchain,
-                                                     failedHeader = header,
-                                                     tipUpdateFailure = err)
-                Future.successful(failed)
-
-              case Some(prevBlockHeader) =>
-                //found a header to connect to!
-                logger.debug(
-                  s"Attempting to add new tip=${header.hashBE.hex} with prevhash=${header.previousBlockHashBE.hex} to chain")
-                val tipResultF =
-                  TipValidation.checkNewTip(newPotentialTip = header,
-                                            currentTip = prevBlockHeader,
-                                            blockHeaderDAO = blockHeaderDAO)
-
-                tipResultF.map { tipResult =>
-                  tipResult match {
-                    case TipUpdateResult.Success(headerDb) =>
-                      val newChain =
-                        Blockchain.fromHeaders(headerDb +: blockchain.headers)
-                      BlockchainUpdate.Successful(newChain, headerDb)
-                    case fail: TipUpdateResult.Failure =>
-                      BlockchainUpdate.Failed(blockchain, header, fail)
-                  }
-                }
+    val tipResultF: Future[BlockchainUpdate] = {
+      val nested: Vector[Future[BlockchainUpdate]] = blockchains.map {
+        blockchain =>
+          val prevBlockHeaderIdxOpt =
+            blockchain.headers.zipWithIndex.find {
+              case (headerDb, _) =>
```

Comment:
Index is not used here right? It is not too cheap an operation to add them so we probably shouldn't.

Is there any way to know if find is going from genesis to tip or tip to genesis? If we can make it do the former we should right?

Reply:
Don't we want to do tip to genesis? That is what we are doing currently. We are more likely to have a reorg closer to the tip.

Reply:
the index is also used below